PresentData/PresentMonTraceConsumer.cpp

```diff
@@ -1139,12 +1181,152 @@ void PMTraceConsumer::HandleDWMEvent(EVENT_RECORD* pEventRecord)
+    }
+
+    if (p->DriverBatchThreadId != 0)
+    {
+        auto batchThreadEventIter = mPresentByThreadId.find(p->DriverBatchThreadId);
+        if (batchThreadEventIter != mPresentByThreadId.end() && batchThreadEventIter->second == p) {
+            mPresentByThreadId.erase(batchThreadEventIter);
+        }
+    }
+
+    // mPresentsBySubmitSequence
+    if (p->QueueSubmitSequence != 0) {
+        auto eventIter = mPresentsBySubmitSequence.find(p->QueueSubmitSequence);
+        if (eventIter != mPresentsBySubmitSequence.end() && (eventIter->second == p)) {
+            mPresentsBySubmitSequence.erase(eventIter);
+        }
+    }
+
+    // mWin32KPresentHistoryTokens
+    if (p->CompositionSurfaceLuid != 0) {
+        PMTraceConsumer::Win32KPresentHistoryTokenKey key(
+            p->CompositionSurfaceLuid,
+            p->Win32KPresentCount,
+            p->Win32KBindId
+        );
+
+        auto eventIter = mWin32KPresentHistoryTokens.find(key);
+        if (eventIter != mWin32KPresentHistoryTokens.end() && (eventIter->second == p)) {
+            mWin32KPresentHistoryTokens.erase(eventIter);
+        }
+    }
+
+    // mDxgKrnlPresentHistoryTokens
+    if (p->TokenPtr != 0) {
+        auto eventIter = mDxgKrnlPresentHistoryTokens.find(p->TokenPtr);
+        if (eventIter != mDxgKrnlPresentHistoryTokens.end() && eventIter->second == p) {
+            mDxgKrnlPresentHistoryTokens.erase(eventIter);
+        }
+    }
+
+    // mBltsByDxgContext
+    if (p->DxgKrnlHContext != 0) {
+        auto eventIter = mBltsByDxgContext.find(p->DxgKrnlHContext);
+        if (eventIter != mBltsByDxgContext.end() && eventIter->second == p) {
+            mBltsByDxgContext.erase(eventIter);
+        }
+    }
+
+    // mLastWindowPresent
+    // 0 is a invalid hwnd
+    if (p->Hwnd != 0) {
+        auto eventIter = mLastWindowPresent.find(p->Hwnd);
+        if (eventIter != mLastWindowPresent.end() && eventIter->second == p) {
+            mLastWindowPresent.erase(eventIter);
+        }
+    }
+
+    // mPresentsWaitingForDWM
+    if (p->PresentInDwmWaitingStruct) {
+        for (auto presentIter = mPresentsWaitingForDWM.begin(); presentIter != mPresentsWaitingForDWM.end(); presentIter++) {
+            // This loop should in theory be short because the present is old.
+            // If we are in this loop for dozens of times, something is likely wrong.
+            if (p == *presentIter)
+            {
+                mPresentsWaitingForDWM.erase(presentIter);
+                break;
+            }
+        }
+    }
+
+    // mPresentsByLegacyBlitToken
+    // LegacyTokenData cannot be 0 if it's in mPresentsByLegacyBlitToken list.
+    if (p->LegacyBlitTokenData != 0) {
+        auto eventIter = mPresentsByLegacyBlitToken.find(p->LegacyBlitTokenData);
+        if (eventIter != mPresentsByLegacyBlitToken.end() && eventIter->second == p) {
+            mPresentsByLegacyBlitToken.erase(eventIter);
+        }
+    }
+}
+

```

Comment:
I don't see any interaction with the existing `HandleStuckPresent()`.  The old concept of a stuck present needs to be ported over... and I think that just means calling RemoveLostPresent() instead.

Reply:
Replaced HandleStuckPresent() with RemoveLostPresent()

Reply:
Heads up that replacing this in HandleDxgkBlt regressed Vulkan Windowed mode tracking, and replacing this in HandleDxgkFlip regressed Full Screen and Borderless Windowed Mode tracking. In both cases no presents are reported.

I think the issue here is that RuntimePresentStop isn't being called to remove these presents from mPresentByThreadId, so when the next  HandleDxgkBlt/HandleDxgkFlip call happens before the previous present has been completed (and thus removed from mPresentByThreadId), FindOrCreatePresent finds that previous present and thinks it's stuck.

Will likely solve this via keeping the HandleStuckPresent logic for Runtime::Other in my next PR.

@jenatali as FYI & check to see whether this seems right.

Reply:
That sounds right: in the VK case (or any without instrumented present API) it's essentially just a stuck "mPresentByThreadId" but we don't want to remove the whole present.

Is the right solution to not add to mPresentByThreadId if the present was created in HandleDxgkBlit/Flip()?

Reply:
That sounds like the right fix to me.

Reply:
Is the interaction with this part of FindOrCreatePresent a concern?
https://github.com/GameTechDev/PresentMon/blob/d1749c3f6711b27a4308eb1ff9f763238b5df08f/PresentData/PresentMonTraceConsumer.cpp#L1418

```
        // If not, lookup the oldest in-progress present that was created by
        // this process but still doesn't have a known PresentMode.  This is
        // the mechanism for DXGK/Win32K events for looking up a present
        // created by DXGI/D3D Present on a different thread. It assumes that
        // batched presents are popped off the front of the driver queue by
        // process in order.
```

I believes this means that the oldest in-progress present needs to have a known PresentMode in Vulkan by the point HandleDxgkBlt and HandleDxgkFlip calls FindOrCreatePresent. Is this safe to assume?

Reply:
In the case of Runtime::Other, there won't have been a present created yet. The Blt or Flip would be the first event you see in those cases.

Reply:
Not inserting the mPresentByThread in DxgBlit and DxgKrnl events didn't end up working, I think partially because Microsoft_Windows_DxgKrnl::Present_Info::Id: relies on mPresentsByThreadId to find it? This solution is a bit complex to reason about.

What about just removing non-D3D9/DXGI runtime presents from mPresentsByThreadId at the end of the Microsoft_Windows_DxgKrnl::Present_Info::Id case? This seems intuitive, and passes Vulkan tests without making exceptions for DxgBlit and DxgKrnl.

Reply:
That sound reasonable. You could even take it a step further and remove it from `mPresentsByThreadId` if either:
* The `Present_Info` event thread ID matches the driver thread ID
* The `Present_Info` event thread ID matches the primary thread ID, and the runtime is `Other`

In both of these cases, `Present_Info` is the last event we'd expect to see from the thread that called it.

Reply:
I think Microsoft_Windows_DxgKrnl::Present_Info::Id makes sense since mPresentsByThreadId's lifetime should correspond with as much as the runtime Present API lifetime as possible. In DXGI/D3D9 that would be RuntimePresentStop, in other runtimes, according to the comment, Present_Info should be the best stopping point.

```
    case Microsoft_Windows_DxgKrnl::Present_Info::Id:
    {
        // This event is emitted at the end of the kernel present, before returning.
        // The presence of this event is used with blt presents to indicate that no
        // PHT is to be expected.
```

Reply:
> That sound reasonable. You could even take it a step further and remove it from `mPresentsByThreadId` if either:
> 
> * The `Present_Info` event thread ID matches the driver thread ID
> * The `Present_Info` event thread ID matches the primary thread ID, and the runtime is `Other`
> 
> In both of these cases, `Present_Info` is the last event we'd expect to see from the thread that called it.

Interesting, does RuntimePresentStop not happen after Present_Info in the first case?

Reply:
It will (or already has), but not on the `Present_info` thread, it's on the same thread that did the DXGI/D3D9 start event.

Reply:
I went with https://github.com/GameTechDev/PresentMon/blob/41e50b9a86d43fe182bbf24cab0691b1f1f545f5/PresentData/PresentMonTraceConsumer.cpp#L700-L702

The thinking is that:
> The Present_Info event thread ID matches the driver thread ID
Is in this code block: https://github.com/GameTechDev/PresentMon/blob/41e50b9a86d43fe182bbf24cab0691b1f1f545f5/PresentData/PresentMonTraceConsumer.cpp#L689-L696

For 
> The Present_Info event thread ID matches the primary thread ID, and the runtime is Other

If runtime is Other, I think we won't get any other events to help remove the presents from mPresentByThreadId, so we should remove altogether.

Not sure if at any point you meant removing the DriverBatchThreadId member, for now they are probably not removed until the present completes.

Reply:
That makes sense, I forgot about that block which removes it for batched presents.